pennylane/devices/tests/test_measurements.py

```diff
@@ -698,6 +968,50 @@ def circuit():
 
         assert np.allclose(res, expected, atol=tol(dev.shots))
 
+    def test_var_projector(self, device, tol):
+        """Tests if the samples of a Projector observable returned by sample have
+        the correct values
+        """
+        n_wires = 2
+        dev = device(n_wires)
+
+        if "Projector" not in dev.observables:
+            pytest.skip("Skipped because device does not support the Projector observable.")
+
+        phi = 0.543
+        theta = 0.654
+
+        @qml.qnode(dev)
+        def circuit(basis_state):
+            qml.RX(phi, wires=[0])
+            qml.RY(theta, wires=[1])
+            qml.CNOT(wires=[0, 1])
+            return qml.var(qml.Projector(basis_state, wires=[0, 1]))
+
+        res = circuit([0, 0])
+        expected = (np.cos(phi / 2) * np.cos(theta / 2)) ** 2 - (
+            (np.cos(phi / 2) * np.cos(theta / 2)) ** 2
+        ) ** 2
+        assert np.allclose(res, expected, atol=tol(dev.shots))
```

Comment:
Amazing test!

Reply:
This makes me realize though, that

```
var(|b><b|) = E(|b><b|b><b) - E(|b><b|)^2
            = E(|b><b) - E(|b><b|)^2
            = marginal_prob[b] - marginal_prob[b]^2
            = marginal_prob[b] (1-marginal_prob[b])
```

So there is a potential to also speed up `QubitDevice.var`, similarly to how you did with `QubitDevice.expval`:

```python
def var(self, observable, shot_range=None, bin_size=None):
    if observable.name == "Projector":
        # branch specifically to handle the projector observable
        idx = int("".join(str(i) for i in observable.parameters[0]), 2)
        probs = self.probability(
            wires=observable.wires, shot_range=shot_range, bin_size=bin_size
        )
        return probs[idx] - probs[idx] ** 2

    if self.shots is None:
        # exact variance value
        eigvals = self._asarray(observable.eigvals, dtype=self.R_DTYPE)
        prob = self.probability(wires=observable.wires)
        return self._dot((eigvals ** 2), prob) - self._dot(eigvals, prob) ** 2

    # estimate the variance
    samples = self.sample(observable, shot_range=shot_range, bin_size=bin_size)
    return np.squeeze(np.var(samples, axis=0))
```

Having said that, I think you have done more than enough for this PR to be merge-ready, so we could simply make a new GitHub issue, and leave this change for a new PR :)

Reply:
This also gives me another thought - we should (in another PR!) add to the device API so that you can do

```python
self.basis_state_probability(basis_state, wires)
```

This might be much more memory efficient for some devices, in that the full probability vector is avoided, since we only require a single element of it.

Reply:
Interesting, I agree with that! It does sound like something to do in the next step to be more memory efficient.